fl4health/clients/mkmmd_clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,274 @@
+        self.init_global_feature_extractor: FeatureExtractorBuffer
+
+    def setup_client(self, config: Config) -> None:
+        super().setup_client(config)
+        self.local_feature_extractor = FeatureExtractorBuffer(
+            model=self.model,
+            flatten_feature_extraction_layers=self.flatten_feature_extraction_layers,
+        )
+
+    def update_before_train(self, current_server_round: int) -> None:
+        super().update_before_train(current_server_round)
+        assert isinstance(self.global_model, nn.Module)
+        # Register hooks to extract features from the local model if not already registered
+        self.local_feature_extractor._maybe_register_hooks()
+        # Clone and freeze the initial weights GLOBAL MODEL. These are used to form the Ditto local
+        # update penalty term.
+        self.init_global_model = self.clone_and_freeze_model(self.global_model)
+        self.init_global_feature_extractor = FeatureExtractorBuffer(
+            model=self.init_global_model,
+            flatten_feature_extraction_layers=self.flatten_feature_extraction_layers,
+        )
+        # Register hooks to extract features from the init global model if not already registered
+        self.init_global_feature_extractor._maybe_register_hooks()
+
+    def _should_optimize_betas(self, step: int) -> bool:
+        assert self.beta_global_update_interval is not None
+        step_at_interval = (step - 1) % self.beta_global_update_interval == 0
+        valid_components_present = self.init_global_model is not None
+        return step_at_interval and valid_components_present
+
+    def update_after_step(self, step: int) -> None:
+        if self.beta_global_update_interval > 0 and self._should_optimize_betas(step):
+            assert self.init_global_model is not None
+            # Get the feature distribution of the local and init global features with evaluation mode
+            local_distributions, init_global_distributions = self.update_buffers(self.model, self.init_global_model)
+            # Update betas for the MK-MMD loss based on gathered features during training
+            if self.mkmmd_loss_weight != 0:
+                for layer in self.flatten_feature_extraction_layers.keys():
+                    self.mkmmd_losses[layer].betas = self.mkmmd_losses[layer].optimize_betas(
+                        X=local_distributions[layer], Y=init_global_distributions[layer], lambda_m=1e-5
+                    )
+
+        return super().update_after_step(step)
+
+    def update_buffers(
+        self, local_model: torch.nn.Module, init_global_model: torch.nn.Module
+    ) -> Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]:
+        """Update the feature buffer of the local and global features."""
+
+        self.local_feature_extractor.clear_buffers()
+        self.init_global_feature_extractor.clear_buffers()
+
+        self.local_feature_extractor.enable_accumulating_features()
```

Comment:
I disable it because of memory issues we might get, if we accumulate always. Because model hooks are applied and called whenever model gets an input, if we don't disable it, feature going to keep accumulate.

Reply:
Ah I think I see. If you don't disable accumulation, there will be accumulation during training as well. I was thinking that you clear the buffers immediately after filling them, so no need to worry about it, but that's not the case.